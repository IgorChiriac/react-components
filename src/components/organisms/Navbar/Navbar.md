### Summary

`<Navbar />` doesn't render anything: it's just a namespace for `<Navbar.Dropdwn />`, `<Navbar.Dropdown />` and `<Navbar.Link />`.

In order to create a navigation bar to be used within a **Zopa application**, you will need to leverage these components together.

### Examples

- Static

```js { "props": { "style": { "transform": "translate3d(0, 0, 0)", "border": "2px solid #efefef" } } }
import { Navbar, Link, HamburgerIcon, ProfileIcon, colors } from '@zopauk/react-components';

<Navbar.Layout
  backgroundColor={colors.base.primary}
  left={
    <Navbar.Dropdown
      id="unique-id"
      ariaLabel="Awesome navbar dropdown"
      items={[{ label: 'one', href: '#' }, { label: 'two', href: '#' }, { label: 'three', href: '#' }]}
      renderOpener={({ open, getOpenerProps }) => (
        <div style={{ padding: 8 }}>
          <Navbar.Link href="#" open={open} {...getOpenerProps()}>
            <HamburgerIcon size="30px" activeColor={colors.neutral.white} inactiveColor={colors.neutral.white} />
          </Navbar.Link>
        </div>
      )}
      renderItem={({ item: { label, href }, getItemProps }) => (
        <Link href={href} {...getItemProps()}>
          {label}
        </Link>
      )}
    />
  }
  right={
    <Navbar.Dropdown
      id="another-unique-id"
      ariaLabel="Awesome navbar dropdown"
      items={[{ label: 'one', href: '#' }, { label: 'two', href: '#' }]}
      renderOpener={({ open, getOpenerProps }) => {
        return (
          <div style={{ padding: 8 }}>
            <Navbar.Link color="#fff" href="#" open={open} {...getOpenerProps()}>
              <ProfileIcon activeColor="#fff" inactiveColor="#fff" />
            </Navbar.Link>
          </div>
        );
      }}
      renderItem={({ item: { label, href }, getItemProps }) => (
        <Link href={href} {...getItemProps()}>
          {label}
        </Link>
      )}
    />
  }
/>;
```

- Responsive

```js { "props": { "style": { "transform": "translate3d(0, 0, 0)", "border": "2px solid #efefef" } } }
import { Navbar, Link, HamburgerIcon, FlexCol, colors } from '@zopauk/react-components';

class ResponsiveNavbar extends React.Component {
  constructor(props) {
    super(props);
    this.renderLeft = this.renderLeft.bind(this);
  }

  renderLeft() {
    const items = [{ label: 'one', href: '#' }, { label: 'two', href: '#' }, { label: 'three', href: '#' }];
    return (
      <div>
        <FlexCol xs="auto" m="hidden">
          <Navbar.Dropdown
            id="yet-another-unique-id"
            ariaLabel="Awesome navbar dropdown"
            items={items}
            renderOpener={({ open, getOpenerProps }) => (
              <div style={{ padding: 8 }}>
                <Navbar.Link color={colors.neutral.white} href="#" open={open} {...getOpenerProps()}>
                  <HamburgerIcon size="30px" activeColor={colors.neutral.white} inactiveColor={colors.neutral.white} />
                </Navbar.Link>
              </div>
            )}
            renderItem={({ item: { label, href }, getItemProps }) => (
              <Link href={href} {...getItemProps()}>
                {label}
              </Link>
            )}
          />
        </FlexCol>
        <FlexCol xs="hidden" m="auto">
          {items.map(({ label, href }, index) => (
            <Link key={`desktop-link-${index}`} color={colors.neutral.white} href={href} style={{ padding: 8 }}>
              {label}
            </Link>
          ))}
        </FlexCol>
      </div>
    );
  }

  render() {
    return <Navbar.Layout backgroundColor="#00B9A7" left={this.renderLeft()} />;
  }
}

<ResponsiveNavbar />;
```
