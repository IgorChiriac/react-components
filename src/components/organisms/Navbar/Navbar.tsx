import React from 'react';

import NavbarDropdown from './NavbarDropdown/NavbarDropdown';
import NavbarLayout from './NavbarLayout/NavbarLayout';
import NavbarLink from './NavbarLink/NavbarLink';

class Navbar extends React.Component {
  public static Layout = NavbarLayout;
  public static Dropdown = NavbarDropdown;
  public static Link = NavbarLink;

  public render() {
    return null;
  }
}

export default Navbar;
