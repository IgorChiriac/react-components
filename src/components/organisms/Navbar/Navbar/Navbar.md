### Summary

Navbar will render an array of links provided using a dropdown which works on click and a link which is a wrap around our [`<Link />`](/#/Components/Atoms/Link) atom, decorated with tow extra props: `active` and `withChevron`. You can opt out displaying the CTA using `withCTA` prop.

⚠️ The menu animates when scrolling. For a better preview view [Default example](#/Components/Organisms/Navbar/Navbar/1)

### Examples

- Default theme

```ts { "props": { "style": { "transform": "translate3d(0, 0, 0)", "backgroundColor": "#FFFFFF", "overflow": "hidden", "height": "500px" } } }
import { Navbar } from '@zopauk/react-components';

const NAV_ITEMS = [
  {
    label: 'Borrow',
    href: '/loans',
    onClick: () => alert('testing2'),
    qadata: 'Borrow.topBar.Button',
    children: [
      {
        label: 'Loans',
        href: '/loans',
        onClick: () => alert('testing'),
        qadata: 'loans.Borrow.topBar.Menu',
      },
      {
        label: 'Car finance',
        groupTitle: 'carFinanceGroup',
        groupChildren: [
          {
            label: 'Car loan',
            href: '/loans/car-loans',
            qadata: 'Car_loan.Borrow.topBar.Menu',
          },
        ],
      },
      {
        label: 'Tools',
        groupTitle: 'toolsGroup',
        groupChildren: [
          {
            label: 'Borrowing power',
            href: '/borrowing-power',
            qadata: 'Borrowing_Power.Borrow.topBar.Menu',
          },
        ],
      },
    ],
  },
  {
    label: 'Invest',
    href: '/invest',
    qadata: 'Invest.topBar.Menu',
    children: [
      {
        label: 'Peer-to peer investments',
        href: '/invest',
        qadata: 'Peer_to_peer.Invest.topBar.Menu',
      },
      {
        label: 'Innovative Finance ISA',
        href: '/invest/isa',
        qadata: 'Innovative.Invest.topBar.Menu',
      },
    ],
  },
  {
    label: 'About',
    href: '/about',
    qadata: 'About.topBar.Menu',
    children: [
      {
        label: 'About Zopa',
        href: '/about',
        qadata: 'About.About.topBar.Menu',
      },
      {
        label: 'Our story',
        href: '/about/our-story',
        qadata: 'Story.About.topBar.Menu',
      },
      {
        label: 'Our board',
        href: '/about/board',
        qadata: 'Our_Board.About.topBar.Menu',
      },
      {
        label: 'Our leadership team',
        href: '/about/leadership',
        qadata: 'Our_Leadership.About.topBar.Menu',
      },
      {
        label: 'Awards',
        href: '/about/awards',
        qadata: 'Awards.About.topBar.Menu',
      },
      {
        label: 'Careers',
        href: '/about/careers',
        qadata: 'Careers.About.topBar.Menu',
      },
      {
        label: 'Press office',
        href: '/about/press',
        qadata: 'Press_office.About.topBar.Menu',
      },
    ],
  },
  {
    label: 'Support',
    href: '/contact',
    qadata: 'Support.topBar.Menu',
  },
];

<Navbar overlayLogoWith={<span></span>} links={NAV_ITEMS} />;
```

- Custom Navbar components

```ts { "props": { "style": { "transform": "translate3d(0, 0, 0)", "backgroundColor": "#FFFFFF", "overflow": "hidden", "height": "500px" } } }
import styled from 'styled-components';

import { Navbar, Link, NavbarLinkStyles } from '@zopauk/react-components';

const NAV_ITEMS = [
  {
    label: 'Borrow',
    href: '/loans',
    qadata: 'Borrow.topBar.Button',
    children: [
      {
        label: 'Loans',
        href: '/loans',
        qadata: 'loans.Borrow.topBar.Menu',
      },
      {
        label: 'Car finance',
        groupTitle: 'carFinanceGroup',
        groupChildren: [
          {
            label: 'Car loan',
            href: '/loans/car-loans',
            qadata: 'Car_loan.Borrow.topBar.Menu',
          },
        ],
      },
      {
        label: 'Tools',
        groupTitle: 'toolsGroup',
        groupChildren: [
          {
            label: 'Borrowing power',
            href: '/borrowing-power',
            qadata: 'Borrowing_Power.Borrow.topBar.Menu',
          },
        ],
      },
    ],
  },
  {
    label: 'Invest',
    href: '/invest',
    qadata: 'Invest.topBar.Menu',
    children: [
      {
        label: 'Peer-to peer investments',
        href: '/invest',
        qadata: 'Peer_to_peer.Invest.topBar.Menu',
      },
      {
        label: 'Innovative Finance ISA',
        href: '/invest/isa',
        qadata: 'Innovative.Invest.topBar.Menu',
      },
    ],
  },
  {
    label: 'About',
    href: '/about',
    qadata: 'About.topBar.Menu',
    children: [
      {
        label: 'About Zopa',
        href: '/about',
        qadata: 'About.About.topBar.Menu',
      },
      {
        label: 'Our story',
        href: '/about/our-story',
        qadata: 'Story.About.topBar.Menu',
      },
      {
        label: 'Our board',
        href: '/about/board',
        qadata: 'Our_Board.About.topBar.Menu',
      },
      {
        label: 'Our leadership team',
        href: '/about/leadership',
        qadata: 'Our_Leadership.About.topBar.Menu',
      },
      {
        label: 'Awards',
        href: '/about/awards',
        qadata: 'Awards.About.topBar.Menu',
      },
      {
        label: 'Careers',
        href: '/about/careers',
        qadata: 'Careers.About.topBar.Menu',
      },
      {
        label: 'Press office',
        href: '/about/press',
        qadata: 'Press_office.About.topBar.Menu',
      },
    ],
  },
  {
    label: 'Support',
    href: '/contact',
    qadata: 'Support.topBar.Menu',
  },
];

// could be a gatsby or react-router-dom <Link />
const SomeLibraryLink = ({ href, ...rest }) => <a href={href} {...rest} />;

const CustomLink = styled(SomeLibraryLink)`
  ${NavbarLinkStyles}
`;

<Navbar
  overlayLogoWith={<a href="https://www.zopa.com" />}
  links={NAV_ITEMS}
  renderLink={(item, index, props) => (
    <CustomLink href={item.href} target="_blank" {...props} className="px-4 py-2" is>
      {item.label}
    </CustomLink>
  )}
/>;
```

- Without CTA

```ts { "props": { "style": { "transform": "translate3d(0, 0, 0)", "backgroundColor": "#FFFFFF", "overflow": "hidden", "height": "500px" } } }
import { Navbar } from '@zopauk/react-components';

const NAV_ITEMS = [
  {
    label: 'Borrow',
    href: '/loans',
    qadata: 'Borrow.topBar.Button',
    children: [
      {
        label: 'Loans',
        href: '/loans',
        qadata: 'loans.Borrow.topBar.Menu',
      },
      {
        label: 'Car finance',
        groupTitle: 'carFinanceGroup',
        groupChildren: [
          {
            label: 'Car loan',
            href: '/loans/car-loans',
            qadata: 'Car_loan.Borrow.topBar.Menu',
          },
        ],
      },
      {
        label: 'Tools',
        groupTitle: 'toolsGroup',
        groupChildren: [
          {
            label: 'Borrowing power',
            href: '/borrowing-power',
            qadata: 'Borrowing_Power.Borrow.topBar.Menu',
          },
        ],
      },
    ],
  },
  {
    label: 'Invest',
    href: '/invest',
    qadata: 'Invest.topBar.Menu',
    children: [
      {
        label: 'Peer-to peer investments',
        href: '/invest',
        qadata: 'Peer_to_peer.Invest.topBar.Menu',
      },
      {
        label: 'Innovative Finance ISA',
        href: '/invest/isa',
        qadata: 'Innovative.Invest.topBar.Menu',
      },
    ],
  },
  {
    label: 'About',
    href: '/about',
    qadata: 'About.topBar.Menu',
    children: [
      {
        label: 'About Zopa',
        href: '/about',
        qadata: 'About.About.topBar.Menu',
      },
      {
        label: 'Our story',
        href: '/about/our-story',
        qadata: 'Story.About.topBar.Menu',
      },
      {
        label: 'Our board',
        href: '/about/board',
        qadata: 'Our_Board.About.topBar.Menu',
      },
      {
        label: 'Our leadership team',
        href: '/about/leadership',
        qadata: 'Our_Leadership.About.topBar.Menu',
      },
      {
        label: 'Awards',
        href: '/about/awards',
        qadata: 'Awards.About.topBar.Menu',
      },
      {
        label: 'Careers',
        href: '/about/careers',
        qadata: 'Careers.About.topBar.Menu',
      },
      {
        label: 'Press office',
        href: '/about/press',
        qadata: 'Press_office.About.topBar.Menu',
      },
    ],
  },
  {
    label: 'Support',
    href: '/contact',
    qadata: 'Support.topBar.Menu',
  },
];

<Navbar withCTA={false} links={NAV_ITEMS} />;
```

- Without links

```ts { "props": { "style": { "transform": "translate3d(0, 0, 0)", "backgroundColor": "#FFFFFF", "overflow": "hidden", "height": "500px" } } }
import { Navbar, Link, buttonStyle } from '@zopauk/react-components';
import styled from 'styled-components';

const ButtonLink = styled(Link)`
  ${buttonStyle}
  margin-left: 8px;
`;

<Navbar
  overlayLogoWith={<a href="https://www.zopa.com" />}
  cta={
    <ButtonLink href="https://www.zopa.com" styling="primary">
      Sign-in
    </ButtonLink>
  }
></Navbar>;
```
