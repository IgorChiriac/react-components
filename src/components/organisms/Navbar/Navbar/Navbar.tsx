import React from 'react';
import styled from 'styled-components';

import { colors } from '../../../../constants/colors';
import { navbarHeight } from '../../../../constants/components';
import useScrollThreshold from '../useScrollThreshold/useScrollThreshold';
import FlexContainer from '../../../layout/FlexContainer/FlexContainer';

export interface ILayoutOuterProps extends React.HTMLAttributes<HTMLDivElement> {
  backgroundColor: string;
  overlap: boolean;
}

const LayoutOuter = styled.div<ILayoutOuterProps>`
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  z-index: 1;
  background-color: ${({ backgroundColor }) => backgroundColor};
  height: ${navbarHeight}px;
  transition: box-shadow 0.2s ease-in-out;
  box-shadow: 0 0 0 transparent;
  ${({ overlap }) => overlap && `box-shadow: rgba(0, 0, 0, 0.2) 0 1px 2px;`}
`;

const LayoutInner = styled.nav`
  display: flex;
  align-items: center;
  justify-content: space-between;
  max-width: 960px;
  height: 100%;
  width: 100%;
  margin-left: auto;
  margin-right: auto;
`;

const Spacer = styled.div`
  height: ${navbarHeight}px;
`;

const Container = styled(FlexContainer)`
  height: inherit;
`;

export interface INavbarProps {
  /**
   * background color
   */
  backgroundColor?: string;
  /**
   * Component appearing on the left
   */
  left?: React.ReactNode;
  /**
   * Component appearing in the center
   */
  center?: React.ReactNode;
  /**
   * Component appearing on the right
   */
  right?: React.ReactNode;
}

const Navbar = ({ backgroundColor = colors.neutral.white, left, center, right }: INavbarProps) => {
  const overThreshold = useScrollThreshold();

  return (
    <>
      <Spacer />
      <LayoutOuter backgroundColor={backgroundColor} overlap={overThreshold}>
        <Container>
          <LayoutInner>
            {left && <div>{left}</div>}
            {center && <div>{center}</div>}
            {right && <div>{right}</div>}
          </LayoutInner>
        </Container>
      </LayoutOuter>
    </>
  );
};

export default Navbar;
