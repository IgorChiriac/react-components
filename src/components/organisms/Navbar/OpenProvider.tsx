import React, { createContext, Dispatch, FC, SetStateAction, useState } from 'react';

interface OpenContextType {
  open: boolean;
  setOpen: Dispatch<SetStateAction<boolean>>;
}

const OpenContext = createContext<OpenContextType>({
  open: false,
  setOpen: () => {
    /* Empty method to enable typing for useState */
  },
});

const OpenProvider: FC = ({ children }) => {
  const [open, setOpen] = useState(false);

  return <OpenContext.Provider value={{ open, setOpen }}>{children}</OpenContext.Provider>;
};

const useOpen = () => React.useContext(OpenContext);

export { OpenProvider, useOpen };
