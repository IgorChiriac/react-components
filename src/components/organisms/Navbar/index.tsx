import React, { FC } from 'react';

import Navbar, { NavbarProps } from './Navbar/Navbar';
import NavbarAction from './NavbarAction/NavbarAction';
import { OpenProvider } from './OpenProvider';

interface NavbarStatic {
  Action: typeof NavbarAction;
}

const NavbarWrapper: NavbarStatic & FC<NavbarProps> = (props) => (
  <OpenProvider>
    <Navbar {...props} />
  </OpenProvider>
);

NavbarWrapper.Action = NavbarAction;

export default NavbarWrapper;
export { navbarLinkStyles } from './NavbarLink/NavbarLink';
