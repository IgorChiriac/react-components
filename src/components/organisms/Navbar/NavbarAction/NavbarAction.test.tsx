import React from 'react';
import { render, screen } from '@testing-library/react';
import Navbar from '..';

describe('<Navbar />', () => {
  it('should render component with default props', () => {
    const { container } = render(<Navbar.Action />);
    expect(container).toMatchSnapshot();
  });

  it('should render component with custom text', () => {
    Object.defineProperty(window, 'innerWidth', { writable: true, configurable: true, value: 1280 });

    render(<Navbar.Action ctaText="My Account" />);
    const button = screen.getByText('My Account');
    expect(button).toBeInTheDocument();
  });
});
