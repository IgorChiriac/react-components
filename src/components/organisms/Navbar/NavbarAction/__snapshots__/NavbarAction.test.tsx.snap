// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Navbar /> should render component with default props 1`] = `
.c0 {
  -webkit-text-decoration: none;
  text-decoration: none;
  box-sizing: border-box;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 16px 24px;
  font-family: "Open Sans",Roboto,Helvetica,Arial,sans-serif;
  font-size: 15px;
  line-height: 1.2;
  font-weight: 600;
  cursor: pointer;
  border-radius: 8px;
  color: #21247F;
  border: 1px solid transparent;
  outline: 0;
  background: transparent;
}

.c0:hover:not(:disabled) {
  background: #EAEBFA;
}

.c0:focus:not(:active) {
  border: 1px solid #FFFFFF;
  box-shadow: 0 0 4px #3B46C4;
}

.c0:disabled {
  cursor: not-allowed;
  background: #F7F7F7;
  color: #818F9B;
}

.c0:active:not(:disabled) {
  border: 1px solid transparent;
  box-shadow: unset;
  opacity: 0.8;
}

@media (max-width:1280px) {
  .c1 {
    padding: 0;
    display: block;
    width: 100%;
    height: 100%;
  }

  .c1:hover:not(:disabled) {
    background: none;
  }
}

@media (min-width:1280px) {
  .c1 {
    margin-left: 8px;
  }
}

<div>
  <button
    class="c0 c1"
  >
    <svg
      aria-hidden="true"
      class="svg-inline--fa fa-user fa-w-14 fa-lg "
      color="#FFFFFF"
      data-icon="user"
      data-prefix="fas"
      focusable="false"
      role="img"
      viewBox="0 0 448 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M224 256c70.7 0 128-57.3 128-128S294.7 0 224 0 96 57.3 96 128s57.3 128 128 128zm89.6 32h-16.7c-22.2 10.2-46.9 16-72.9 16s-50.6-5.8-72.9-16h-16.7C60.2 288 0 348.2 0 422.4V464c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48v-41.6c0-74.2-60.2-134.4-134.4-134.4z"
        fill="currentColor"
      />
    </svg>
  </button>
</div>
`;
